rust/bridge/node/futures/src/context.rs

```diff
@@ -0,0 +1,470 @@
+                match Rc::try_unwrap(unsafe { Pin::into_inner_unchecked(self.shared_state) }) {
+                    Ok(state) => std::mem::drop(state),
+                    Err(shared_state) => {
+                        shared_state.update(|state| state.very_unsafe_current_context = None)
+                    }
+                };
+
+                // Dummy return value; see https://github.com/neon-bindings/neon/issues/630
+                Ok(cx.undefined())
+            })
+        }));
+
+        // Okay, we've finished! But was there an exception or panic?
+        // Neon translates unwinding panics into JavaScript exceptions,
+        // but that doesn't help us in an asynchronous context.
+        // To preserve safety and catch bugs, we print the error and abort.
+        match result {
+            Ok(Ok(_)) => return,
+            Ok(Err(js_err)) => {
+                eprintln!(
+                    "\n!! Thrown errors must be handled during async execution !!\n{}\n",
+                    js_err
+                        .to_string(cx)
+                        .expect("could not stringify thrown error")
+                        .value()
+                );
+            }
+            Err(panic) => {
+                eprintln!(
+                    "\n!! Panics must be handled during async execution !!\n{}\n",
+                    describe_any(&panic)
+                );
+            }
+        }
+        std::process::abort()
+    }
+
+    /// Tracks outstanding futures to make sure there's always something to block on.
+    pub(crate) fn register_future(&self) {
+        self.shared_state
+            .update(|state| state.num_pending_js_futures += 1)
+    }
+
+    /// Notes that a future has been fulfilled.
+    ///
+    /// See also [register_future](fn@Self::register_future).
+    pub(crate) fn fulfill_future(&self) {
+        self.shared_state
+            .update(|state| state.num_pending_js_futures -= 1)
+    }
+
+    /// Returns the unique, pinned address associated with this context.
+    ///
+    /// Used for context data.
+    fn opaque_id(&self) -> *const () {
+        self.shared_state.as_address()
+    }
+
+    /// Initializes a fresh context.
+    fn new() -> Self {
+        let result = Self {
+            shared_state: Rc::pin(UnwindSafetyRefCell::new(JsAsyncContextImpl {
+                very_unsafe_current_context: None,
+                num_pending_js_futures: 0,
+                complete: false,
+                pool: Some(LocalPool::new()),
+                global_key: String::new(), // replaced below based on the address this object gets pinned to
+                next_global_id: 0,
+                _pinned: PhantomPinned,
+            })),
+        };
+        result.shared_state.update(|state| {
+            state.global_key = format!(
+                "__signal_neon_futures::JsAsyncContext({:p})",
+                result.opaque_id()
```

Comment:
The `impl Drop` is on `JsAsyncContextImpl`, since that's the shared state behind `Rc`. (We could rename the public type to JsAsyncContextRef if we want to make it explicit that `clone()` is sharing.)

Reply:
Thanks I had missed that `Drop`